pkg/notifications/notifications.go

```diff
@@ -0,0 +1,428 @@
+			ev.Message = "Downtime expired"
+		}
+
+	default:
+		return nil, fmt.Errorf("unexpected event type %q", h.EventType)
+	}
+
+	return ev, nil
+}
+
+// buildFlappingHistoryEvent from a flapping history entry.
+func (client *Client) buildFlappingHistoryEvent(ctx context.Context, h *v1history.FlappingHistory) (*event.Event, error) {
+	res, err := client.fetchHostServiceName(ctx, h.HostId, h.ServiceId)
+	if err != nil {
+		return nil, err
+	}
+
+	ev, err := client.buildCommonEvent(res)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build event for %q,%q", res.HostName, res.ServiceName)
+	}
+
+	if h.PercentStateChangeEnd.Valid {
+		ev.Type = event.TypeFlappingEnd
+		ev.Message = fmt.Sprintf(
+			"Checkable stopped flapping (Current flapping value %.2f%% < low threshold %.2f%%)",
+			h.PercentStateChangeEnd.Float64, h.FlappingThresholdLow)
+	} else if h.PercentStateChangeStart.Valid {
+		ev.Type = event.TypeFlappingStart
+		ev.Message = fmt.Sprintf(
+			"Checkable started flapping (Current flapping value %.2f%% > high threshold %.2f%%)",
+			h.PercentStateChangeStart.Float64, h.FlappingThresholdHigh)
+		ev.Mute = types.MakeBool(true)
+		ev.MuteReason = "Checkable is flapping"
+	} else {
+		return nil, errors.New("flapping history entry has neither percent_state_change_start nor percent_state_change_end")
+	}
+
+	return ev, nil
+}
+
+// buildAcknowledgementHistoryEvent from an acknowledgment history entry.
+func (client *Client) buildAcknowledgementHistoryEvent(ctx context.Context, h *v1history.AcknowledgementHistory) (*event.Event, error) {
+	res, err := client.fetchHostServiceName(ctx, h.HostId, h.ServiceId)
+	if err != nil {
+		return nil, err
+	}
+
+	ev, err := client.buildCommonEvent(res)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build event for %q,%q", res.HostName, res.ServiceName)
+	}
+
+	if !h.ClearTime.Time().IsZero() {
+		ev.Type = event.TypeAcknowledgementCleared
+		ev.Message = "Checkable was cleared"
+
+		if h.ClearedBy.Valid {
+			ev.Username = h.ClearedBy.String
+		}
+	} else if !h.SetTime.Time().IsZero() {
+		ev.Type = event.TypeAcknowledgementSet
+
+		if h.Comment.Valid {
+			ev.Message = h.Comment.String
+		} else {
+			ev.Message = "Checkable was acknowledged"
+		}
+
+		if h.Author.Valid {
+			ev.Username = h.Author.String
+		}
+	} else {
+		return nil, errors.New("acknowledgment history entry has neither a set_time nor a clear_time")
+	}
+
+	return ev, nil
+}
+
+// Submit this [database.Entity] to the Icinga Notifications API.
+//
+// Based on the entity's type, a different kind of event will be constructed. The event will be sent to the API in a
+// blocking fashion.
+//
+// Returns true if this entity was processed or cannot be processed any further. Returns false if this entity should be
+// retried later.
+//
+// This method usees the Client's logger.
+func (client *Client) Submit(entity database.Entity) bool {
+	client.submissionMutex.Lock()
+	defer client.submissionMutex.Unlock()
```

Comment:
What exactly is this mutex protecting? Like does it really need to span the entire function?